File: 2D Array (matrix)/48. Rotate Image/Rotate_Image.cpp
#include<bits/stdc++.h>
using namespace std;
// Rotate the image by 90 degrees clockwise
class Solution {
    public:
        void rotate(vector<vector<int>>& matrix) {
            
            // Transpose the matrix
            // Swap elements across the diagonal
            for(int i=0; i<matrix.size(); i++){
                for(int j=0; j<i+1; j++){
                    swap(matrix[i][j],matrix[j][i]);
                }
            }
            // Reverse each row of the transposed matrix
            // Swap elements across the middle row
            for(int i=0; i<matrix.size(); i++){
                reverse(matrix[i].begin(),matrix[i].end());
            }
            
    }
};

int main(){
    Solution s;
    vector<vector<int>> matrix = {{1,2,3},{4,5,6},{7,8,9}};
    s.rotate(matrix);
    for(auto i:matrix){
        for(auto j:i){
            cout<<j<<" ";
        }
        cout<<endl;
    }
    return 0;
}

File: 2D Array (matrix)/48. Rotate Image/Rotate_Image.java

class Solution {

    public void rotate(int[][] matrix) {
        int size = matrix.length;

        //travser matrix 
        for (int i = 0; i < size; i++) {
            for (int j = 0; j < i; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        //reverse matrix each row
        for (int i = 0; i < size; i++) {

            int start = 0;
            int end = matrix[i].length - 1;
            while (start < end) {
                int temp = matrix[i][start];
                matrix[i][start] = matrix[i][end];
                matrix[i][end] = temp;
                start++;
                end--;
            }
        }
    }
}

public class Rotate_Image {

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[][] matrix = {
            {1, 2, 3},
            {4, 5, 6},
            {7, 8, 9}
        };

        solution.rotate(matrix);

        // Print the rotated matrix
        for (int[] row : matrix) {
            for (int value : row) {
                System.out.print(value + " ");
            }
            System.out.println();
        }
    }
}
